### PR TITLE
12.0 imp sale min qty

### PR DESCRIPTION
**Pull request by mourad-ehm**
Improve  sale_order_min_qty module (renamed to sale_restricted_qty)  by adding multiple and max qties order management. 
It add hierarchical setting management.
Settings are defined on product, if values are not set we use settings from product template. If settings are not set on product template we use settings from product category otherwise parent category.

### CODE REVIEW

**Comment by mourad-ehm** [comment]
ping @sebastienbeau

**Comment by bealdav** [comment]
Not sure and not mandatory but it seems there is some duplicate code. Did you consider mixin to avoid that ?

**Comment by mourad-ehm** [comment]
@bealdav , 
Thanks for review. 
Code refactored. Could you check ?

File: sale_order_min_qty/models/product_category.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+from odoo import api, fields, models
+from odoo.addons import decimal_precision as dp
+
+
+class ProductCategory(models.Model):
+    _name = "product.category"
+    _inherit = ["product.category", "product.min.multiple.mixin"]
+
+    def _get_sale_min_multiple_qty(self):
+        res = super(ProductCategory, self)._get_sale_min_multiple_qty()
```
Comment (bealdav) [inline sale_order_min_qty/models/product_category.py]: In python 3 no more super with the same class name than current one
```suggestion
        res = super()._get_sale_min_multiple_qty()
```
Several super in this case

File: sale_order_min_qty/models/sale.py [hunk 2]
```diff
@@ -23,47 +23,82 @@ class SaleOrderLine(models.Model):
...
                 float_compare(
-                    line.product_uom_qty,
-                    line.sale_min_qty,
-                    precision_rounding=rounding,
+                    product_qty, line.sale_min_qty, precision_rounding=rounding
                 )
                 < 0
             )
+
+    @api.multi
```
Comment (bealdav) [inline sale_order_min_qty/models/sale.py]: no more api.multi in v13, make migration easier, remove
```suggestion
```

File: sale_order_min_qty/models/product_min_multiple_mixin.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+class ProductMinMultipleMixin(models.AbstractModel):
+    _name = "product.min.multiple.mixin"
+
+    sale_multiple_qty = fields.Float(
+        compute="_compute_sale_min_multiple_qty",
+        store=True,
+        help="Define sale multiple qty"
+        " 'If not set', Odoo will"
+        " use the value defined in the parent category.",
```
Comment (bealdav) [inline sale_order_min_qty/models/product_min_multiple_mixin.py]: Is it always ok for product or template ?

File: sale_order_min_qty/models/sale.py [hunk 4]
```diff
@@ -23,47 +23,82 @@ class SaleOrderLine(models.Model):
...
+            msg += _(
+                "\n* If you want sell quantity not multiple Quantity"
+                ",Set multiple quantity to 0 on product or product template"
+                " or product category"
+            )
+
+        if msg:
+            raise ValidationError(msg)
 
     @api.multi
```
Comment (bealdav) [inline sale_order_min_qty/models/sale.py]:
```suggestion
```

File: sale_order_min_qty/models/product_min_multiple_mixin.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        digits=dp.get_precision("Product Unit of Measure"),
+    )
+    manual_sale_multiple_qty = fields.Float(
+        string="multiple Sale Qty", digits=dp.get_precision("Product Unit of Measure")
```
Comment (bealdav) [inline sale_order_min_qty/models/product_min_multiple_mixin.py]:
```suggestion
        string="Multiple Sale Qty", digits=dp.get_precision("Product Unit of Measure")
```

File: sale_order_min_qty/views/sale_views.xml [hunk 6]
```diff
@@ -18,16 +20,21 @@
...
                <field name="sale_min_qty"/>
                <field name="is_qty_less_min_qty" invisible="1"/>
                <field name="force_sale_min_qty" invisible="1"/>
+               <field name="is_qty_not_multiple_qty" invisible="1"/>
            </xpath>
 
             <xpath expr="//field[@name='order_line']/tree" position="attributes">>
                 <attribute name="decoration-danger">is_qty_less_min_qty ==  True and force_sale_min_qty == False</attribute>
                 <attribute name="decoration-warning">is_qty_less_min_qty ==  True and force_sale_min_qty == True</attribute>
+                <attribute name="decoration-danger">is_qty_not_multiple_qty ==  True</attribute>
```
Comment (bealdav) [inline sale_order_min_qty/views/sale_views.xml]:
```suggestion
                <attribute name="decoration-danger">is_qty_not_multiple_qty == True</attribute>
```

**Review by bealdav — CHANGES_REQUESTED** [review]
It's really easier to read with the mixin. Thanks

**Comment by mourad-ehm** [comment]
I do not understand error of runbot.
Someone can explain me ?

File: sale_order_min_qty/models/product_min_multiple_mixin.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        string="Multiple Sale Qty", digits=dp.get_precision("Product Unit of Measure")
+    )
+    sale_min_qty = fields.Float(
+        compute="_compute_sale_min_multiple_qty",
+        store=True,
+        help="Define sale min qty"
+        " 'If not set, Odoo will"
+        " use the value defined in the parent object."
+        "Hierarchy is in this order :"
+        "Product/product Template/product category/parent categoroies ",
```
Comment (bealdav) [inline sale_order_min_qty/models/product_min_multiple_mixin.py]: Small remarks, not mandatory:
- there are typo errors and expression to fix 
- there are duplicate strings : if you would use CONSTANT VAR to define strings, there will no duplicates and make it easy to fix
- Missing \n make odoo help popup difficult to read

**Comment by bealdav** [comment]
Runbot is accessible when I test

**Comment by bealdav** [comment]
Naming and ergonomy could be improved in product and in category

Selection_011
 i.e.
```
<group name=".." string="Sale Quantity">
   <group>
      <group>
         <field ... 1>
         <field ... 2>
      </group>
      <group>
         <field ... 3>
         <field ... 3>
      </group>
   </group>
</group>

```

With computed at left and manual at right for example ?

**Comment by mourad-ehm** [comment]
Why runbot test is red ? which is the problem ?

**Comment by rousseldenis** [comment]
> Why runbot test is red ? which is the problem ?

You need to click on runbot details. Then, look at the icon (danger or error). Then, click on the cogs and select 'Full all logs'. Depending on icon, look for ERROR or WARNING. You'll find your problem. I rebuild as logs were no more accessible

**Comment by mourad-ehm** [comment]
When I click on runbot detail a got :
image

and full base log display:
image

**Comment by rousseldenis** [comment]
> When I click on runbot detail a got :
> image
> 
> and full base log display:
> image

Ok, will see with Alex. But you can test locally, look for WARNING and the error is there

**Comment by bealdav** [comment]
Click on "Full all logs"

**Comment by mourad-ehm** [comment]
Full all logs give the same error.

**Comment by HaraldPanten** [comment]
Can you force a new push and see if it works, then?

**Comment by ValentinVinagre** [comment]
Any solution for runbot? same problem.

**Comment by mourad-ehm** [comment]
runbot display this errors: ??
Done. Phantomjs binary available at /home/odoo/build/OCA/sale-workflow/node_modules/phantomjs-prebuilt/lib/phantom/bin/phantomjs
[91mnpm WARN saveError ENOENT: no such file or directory, open '/home/odoo/build/OCA/sale-workflow/package.json'
[91mERROR: pylint-odoo 0.0.1.dev459 has requirement docutils==0.14, but you'll have docutils 0.12 which is incompatible.
ERROR: pylint-odoo 0.0.1.dev459 has requirement lxml>=4.2.3, but you'll have lxml 3.7.1 which is incompatible.r ?

**Comment by mourad-ehm** [comment]
Thanks @sebastienbeau for refactor code.
Could you merge this PR Please ?

**Review by sebastienbeau — APPROVED** [review]
LGTM code review and test

**Review by bealdav — APPROVED** [review]

**Comment by gaikaz** [comment]
One of our clients might need this functionality for Odoo 13.0 pretty soon, but this PR is changing the name of the module.
So I suppose it's not a good idea to migrate it as `sale_min_qty`?
Can we push this PR to get merged?

**Comment by rousseldenis** [comment]
@mourad-ehm @sebastienbeau @bealdav This seems a bad idea renaming the module in a same version. Why not keeping just enhancements and proposing renaming to 13 or 14 ?

**Pull request merged**